### PR TITLE
#9 Implemented token based authorization

### PR DESCRIPTION
**Pull request by AndreyShelest**
New web client features:
- creating new user
- log in for existing users
- adding new budget items for authorized users

### CODE REVIEW

File: sources/server/handlers.py [hunk 1]
```diff
@@ -16,35 +16,112 @@
...
-class MainHandler(RequestHandler):
+    def get_current_user(self):
+        user_token = self.get_secure_cookie("user_token")
+        if user_token is None:
+            return None
+
+        return self.session.query(User).filter_by(token=user_token.decode()).first()
+
+    def has_users(self):
+        return self.session.query(User).first() is not None
```
Comment (approximator) [inline sources/server/handlers.py]:
```suggestion
        return self.session.query(User).count()
```

File: .project [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (approximator) [inline .project]: What is this file for?

**Review by approximator — APPROVED** [review]

File: .project (hunk 2 above)
Comment (AndreyShelest) [reply .project]: it was removed in last commit

**Review by Radeon4650 — APPROVED** [review]
- Migration tools needed.
- **Number** field of table outputs item id, not the actual number in current list.
- Needed **Date picker** for inserting new item.
- Needed Table separation by **Category**.

File: sources/db/__init__.py [hunk 3]
```diff
@@ -16,6 +16,24 @@
...
+
+def make_db(db_uri):
+    session_factory = make_session_factory(db_uri)
+    fill_db_if_empty(session_factory.engine)
+    return session_factory
+
+
+def fill_db_if_empty(engine):
+    table_names = sa.inspect(engine).get_table_names()
+    is_empty = table_names == []
```
Comment (approximator) [inline sources/db/__init__.py:36]: This won't work if you add another table to the DB.

I think it's safe to call `BASE_MODEL.metadata.create_all(engine)` without any prior checks.

**Review by approximator — APPROVED** [review]

**Pull request merged**